sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala

```diff
@@ -69,16 +77,16 @@ object DynamicJoinSelection extends Rule[LogicalPlan] {
   }
 
   def apply(plan: LogicalPlan): LogicalPlan = plan.transformDown {
-    case j @ ExtractEquiJoinKeys(_, _, _, _, _, left, right, hint) =>
+    case j @ ExtractEquiJoinKeys(joinType, _, _, _, _, left, right, hint) =>
       var newHint = hint
       if (!hint.leftHint.exists(_.strategy.isDefined)) {
-        selectJoinStrategy(left).foreach { strategy =>
+        selectJoinStrategy(left, joinType).foreach { strategy =>
```

Comment:
For left outer join, and the left side has many empty partitions, shall we still demote?

Reply:
For LOJ with many empty partitions on the left, the local join can short-circuit whether you broadcast or shuffle.  I'm not sure how to determine which strategy will send less data around.  Is there another heuristic that can be used?

Reply:
> the local join can short-circuit whether you broadcast or shuffle.

I think there is a misunderstanding here. We see many empty partitions on the shuffled left side, it doesn't mean the original left side before shuffle also has many empty partitions. I think we need to demote.

@ekoifman can you open a followup PR to fix this issue?

Reply:
You're right.  filed SPARK-37753